### PR TITLE
Layout and commit distance data

### LINKED ISSUE #3: Show commit distance alongside age of build
The widget shows the age of the build but when there isn't much churn this isn't so useful. We should also show the commit distance (and use that as part of evaluating if a dependency is "good")

### PR DESCRIPTION
**Pull request by analogrelay**
Fixes https://github.com/Pilchie/DependencyFlow/issues/3

**First, the glamour shot:**

Not sure why the other widgets weren't loading 🤷‍♂. Also, I fixed the contrast for "yellow" tiles already...

image

1. I added a front page that lets you pick a channel and repo, then you get a URL you can re-use. This allows you to generate this widget for any repo. Once you do that, you can take the URL in the address bar and embed it/link it/etc. at your leisure. For example, on `dotnet/core-sdk`:

image

2. I needed some more gnarly(er) code to compute C O O L metrics so I moved the logic out of the razor. I don't mind how we do it, moving that to C# was just a little easier now that there was some extra logic there (Razor just doesn't have the same refactoring support yet 😢)

3. I computed the "condition" of the dependency based on the following (pulled-from-nowhere) heuristic:
   * A dependency is ✔️GOOD✔️ if the build we have is <= 3 days old **or** we are less than 25 commits behind. The "or" is to cover cases where there isn't much churn. 25 commits is just a number I made up.
   * A dependency is ⚠️CAUTION⚠️ if the build we have is > 3 and <= 7 days old **and** we are more than 25 commits behind.
   * A dependency is ❌DANGER❌ if the build we have is > 7 days old **and** we are more than 25 commits behind.

4. I added a dark theme because I'm a 10x engineer... and to blend in with AzDO. It's enabled just by setting `?theme=dark` in the query string. Super sketchy, but it works :)

5. I used Bootstrap's [card decks](https://getbootstrap.com/docs/4.3/components/card/#card-decks) to get a fancy layout going. It seemed easiest and snazziest.

Computing the commit distance isn't fast unfortunately (though only takes a few seconds, so should be reasonable for the dashboard) and requires a GitHub token be provided (it doesn't *require* it but you'll blow through the anonymous rate limit pretty quick).

cc @davidfowl @rynowak

### CODE REVIEW

**Comment by analogrelay** [comment]
I fixed the contrast for "yellow" tiles in the dark theme (visible in the AzDO screenshot):

image

File: Pages/Incoming.cshtml.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            _github = github;
+        }
+
+        public IReadOnlyList<IncomingRepo> IncomingRepositories { get; private set; }
+        public RateLimit CurrentRateLimit { get; private set; }
+
+        public async Task OnGet(int channelId, string repo)
+        {
+            // The 'repo' is URL encoded (because it's a URL within a URL). Decode it before calling the API.
+            var decodedRepo = WebUtility.UrlDecode(repo);
```
Comment (rynowak) [inline Pages/Incoming.cshtml.cs:29]: We're hoping to add an attribute that does this for you in 5.0 😆

File: Pages/Shared/_Layout.cshtml [hunk 2]
```diff
@@ -1,4 +1,10 @@
-﻿<!DOCTYPE html>
+﻿@{ 
+    var darkMode = string.Equals(Context.Request.Query["theme"], "dark")
```
Comment (rynowak) [inline Pages/Shared/_Layout.cshtml:2]: nice

Comment (analogrelay) [reply Pages/Shared/_Layout.cshtml:2]: Nathan Fillian trying to tell someone not to disclose a secret by waving his hand and putting a finger up to his mouth in a "shh" gesture

Dark Theme Easy Mode

File: Pages/Incoming.cshtml.cs (hunk 1 above)
Comment (analogrelay) [reply Pages/Incoming.cshtml.cs:29]: I'm 100% fine with doing this, but an attribute would be niceness

Comment (rynowak) [reply Pages/Incoming.cshtml.cs:29]: Yeah, it's a case of old school vs new school.

The fact that we don't/can't know when to unescape slashes for you catches people out. If you have a feature for it, then it seems like we're setting users up for success.  It comes up A LOT.

**Review by Pilchie — APPROVED** [review]
Not sure why I didn't get any GitHub notifications for this, but looks good to me.

**Comment by analogrelay** [comment]
I think you were OOF when I made it. Perhaps it happened but got lost :).

**Comment by Pilchie** [comment]
You should have pestered me.

**Comment by analogrelay** [comment]
Touché

**Comment by Pilchie** [comment]
Well, I borked it.  Seems I have a Preview 8 build on my int.Preview and don't have the right feed to be able to restore.  Will work on fixing this morning.

**Comment by Pilchie** [comment]
Nitpicking - is there any way to pick up the theme from the host page?  Not all of us are dark theme fans...

**Comment by analogrelay** [comment]
Not trivially. The `iframe` is sandboxed and protected from accessing anything from the root page. Honestly, the white background works OK on the dark theme for now, so if you want to change it back that's OK with me.

**Comment by analogrelay** [comment]
An option that would work is to store theme settings in `window.localStorage` or a cookie. Then the option is per "user". I'll file a bug for that

**Comment by analogrelay** [comment]
Or rather, update #8 with context ;)

**Pull request merged**